contracts/PoolRouter.sol

```diff
@@ -40,43 +40,51 @@ contract PoolRouter {
     function batch(
         PoolDataTypes.Operation[] calldata operations,
         bytes[] calldata data
-    ) external payable {
+    ) external payable returns(bytes[] memory resultValues) {
         require(operations.length == data.length, "Mismatched operation data");
         PoolAddresses memory cache;
 
+        resultValues = new bytes[](operations.length);
+
         for (uint256 i = 0; i < operations.length; i += 1) {
             PoolDataTypes.Operation operation = operations[i];
             
             if (operation == PoolDataTypes.Operation.ROUTE) {
                 (address base, address fyToken, bytes memory poolcall) = abi.decode(data[i], (address, address, bytes));
                 if (cache.base != base || cache.fyToken != fyToken) cache = PoolAddresses(base, fyToken, findPool(base, fyToken));
-                _route(cache, poolcall);
+                (bool success, bytes memory result) = _route(cache, poolcall);
```

Comment:
Can we use `results[i] = abi.encode(action(...))` throughout? It's compact, yet readable.

Reply:
I know! Unfortunately, it is not possible Solidity-wise, I've spent some time researching that yesterday.